File: Ballast/AppModel.swift
import CallKit
import Foundation
import SwiftUI
import UIKit
import UserNotifications

enum Route {
    case setup
    case session
    case asking
    case closed
}

@Observable
@MainActor
final class AppModel {
    private(set) var state = SessionState()
    private(set) var route: Route = .setup
    private(set) var closed: SessionRecord?

    /// Peak |Δ| per 110 ms bucket. 160 samples ≈ 18 seconds.
    private(set) var trace: [Double] = []
    private(set) var askedAt: TimeInterval = 0
    private(set) var nowTick: TimeInterval = Date().timeIntervalSince1970

    var draft: Prefs

    let store: Store
    let detector = MotionDetector()
    let liveActivity = LiveActivityController()
    let notifications = Notifications()
    let screenTime = ScreenTime()

    private var timer: Timer?
    private let callObserver = CXCallObserver()
    private var callDelegate: CallWatcher?
    private var lastEngineTick: TimeInterval = 0
    private var bucketPeak: Double = 0
    private var bucketStart: TimeInterval = 0
    private var awaySince: TimeInterval?

    /// Total time the app spent out of the foreground during this session — time iOS
    /// gave us no motion at all. Surfaced rather than hidden.
    private(set) var awayTotal: TimeInterval = 0

    init(store: Store = Store()) {
        self.store = store
        self.draft = store.prefs

        detector.onPickup = { [weak self] t in self?.send(.pickup, at: t) }
        detector.onSample = { [weak self] magnitude in self?.sample(magnitude) }

        let watcher = CallWatcher { [weak self] active in
            self?.send(.callChanged(active))
        }
        callDelegate = watcher
        callObserver.setDelegate(watcher, queue: .main)

        UNUserNotificationCenter.current().delegate = notifications
        // Tapping the question is a pick-up: let the engine decide, exactly as it
        // does for a real one.
        notifications.onOpen = { [weak self] in self?.send(.pickup) }

        // After resume: draining first would clear the timestamp with no session
        // to receive it.
        if let open = store.open {
            resume(open)
            drainMonitorSlip()
        }
    }

    // MARK: - Derived

    var config: Config {
        Config(
            intervalMin: draft.intervalMin, mode: draft.mode,
            holdSeconds: draft.holdSeconds, sensitivity: draft.sensitivity)
    }

    var phase: Phase { state.phase(at: nowTick) }

    var elapsed: TimeInterval { nowTick - (state.sessionStart ?? nowTick) }

    /// Seconds until the next pick-up would ask. Nil when nothing is owed.
    var armedIn: TimeInterval? {
        guard let anchor = state.anchor else { return nil }
        return max(0, state.config.interval - (nowTick - anchor))
    }

    var hold: Int {
        holdRemaining(
            holdSeconds: state.config.holdSeconds, askedAt: askedAt, now: nowTick)
    }

    /// Minutes since the anchor, for the asking copy.
    var minutesSinceAnchor: Int {
        guard let anchor = askAnchor else { return draft.intervalMin }
        return Int((askedAt - anchor) / 60)
    }

    private var askAnchor: TimeInterval?

    var open: SessionRecord? { store.open }

    // MARK: - Intent

    func start() {
        guard !draft.task.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        store.save(prefs: draft)
        let now = Date().timeIntervalSince1970
        store.begin(task: draft.task, at: now)
        trace = []
        closed = nil
        awayTotal = 0
        awaySince = nil
        detector.threshold = Constants.sensitivityLadder[draft.sensitivity - 1]
        detector.start()
        route = .session
        startClock()
        UIApplication.shared.isIdleTimerDisabled = true
        send(.startSession(task: draft.task, config: config), at: now)
        liveActivity.start(task: draft.task, armedAt: armedAtDate, endsAt: endsAtDate)
        syncActivity()
        armNotifications()

        SharedState.task = draft.task
        SharedState.intervalMin = draft.intervalMin
        SharedState.endsAt = endsAtDate?.timeIntervalSince1970 ?? 0
        SharedState.sessionActive = true
        screenTime.startMonitoring()
    }

    func end() {
        let now = Date().timeIntervalSince1970
        send(.endSession, at: now)
        closed = store.end(at: now)
        detector.stop()
        stopClock()
        UIApplication.shared.isIdleTimerDisabled = false
        liveActivity.end()
        notifications.cancel()
        SharedState.sessionActive = false
        screenTime.stopMonitoring()
    }

    /// iOS suspends accelerometer delivery outside the foreground, so a session that
    /// leaves the screen is a session that sees nothing. Keep the screen awake while
    /// one is running, and account for any gap when the app comes back.
    func setForeground(_ active: Bool) {
        guard state.sessionStart != nil else {
            UIApplication.shared.isIdleTimerDisabled = false
            return
        }
        if active {
            drainMonitorSlip()
            if let since = awaySince { awayTotal += Date().timeIntervalSince1970 - since }
            awaySince = nil
            UIApplication.shared.isIdleTimerDisabled = true
            detector.start()
            startClock()
        } else {
            awaySince = Date().timeIntervalSince1970
            UIApplication.shared.isIdleTimerDisabled = false
            detector.stop()
            stopClock()
        }
    }

    func logSlip() { send(.slip) }

    /// The simulator has no accelerometer, and the spec puts this under the slip button
    /// regardless. It goes through the same engine path a real pick-up does.
    func simulatePickup() { send(.pickup) }

    func dismiss(_ how: Dismissal) { send(.dismiss(how)) }

    func newSession() {
        closed = nil
        route = .setup
    }

    // MARK: - Engine

    private func send(_ event: Event, at now: TimeInterval = Date().timeIntervalSince1970) {
        // Slip mode consumes the anchor when the question fires, so the copy's "n
        // minutes ago" has to be read before the reduce, not after.
        let anchorBefore = state.anchor
        let (next, effects) = reduce(state, event, now)
        state = next
        for effect in effects { apply(effect, at: now, anchorBefore: anchorBefore) }
        if case .tick = event {} else { syncActivity() }
    }

    /// When the work is due. Held on the store so it survives process death.
    var endsAtDate: Date? {
        guard let start = state.sessionStart else { return nil }
        return Date(timeIntervalSince1970: start + Double(draft.durationMin) * 60)
    }

    /// Minutes left on the work, for the reminder copy.
    var minutesLeft: Int {
        guard let endsAt = endsAtDate else { return 0 }
        return Int((endsAt.timeIntervalSince1970 - nowTick) / 60)
    }

    /// When the next pick-up starts asking, as a wall-clock date for the Lock Screen.
    private var armedAtDate: Date? {
        state.anchor.map { Date(timeIntervalSince1970: $0 + state.config.interval) }
    }

    private func syncActivity() {
        guard state.sessionStart != nil else { return }
        liveActivity.update(armedAt: armedAtDate, asking: state.isAsking, endsAt: endsAtDate)
        rescheduleQuestions()
    }

    /// The app is asleep when the questions come due, so the whole series is scheduled
    /// ahead and rebuilt whenever the anchor moves.
    private func rescheduleQuestions() {
        guard let start = state.sessionStart else {
            notifications.cancel()
            return
        }
        // Reminders are about the work, not about a slip. Running them from the
        // engine's anchor meant "After a scroll" sessions — where the anchor is nil
        // until a feed is logged — scheduled nothing at all.
        let base = max(state.lastNudge ?? start, start)
        notifications.schedule(
            task: state.task,
            from: Date(timeIntervalSince1970: base),
            every: state.config.interval,
            endsAt: endsAtDate)
    }

    private func apply(_ effect: Effect, at now: TimeInterval, anchorBefore: TimeInterval?) {
        switch effect {
        case let .record(kind, t):
            store.append(kind, at: t)
        case let .present(screen):
            switch screen {
            case .asking:
                askAnchor = anchorBefore
                askedAt = now
                nowTick = now
                route = .asking
            case .closed:
                route = .closed
            }
        case .dismissAsking:
            route = .session
        }
    }

    private func resume(_ record: SessionRecord) {
        var s = SessionState(task: record.task, config: config, sessionStart: record.startedAt)
        for event in record.events {
            switch event.type {
            case .slip: s.lastSlip = event.t
            case .pickup: s.lastPickup = event.t
            case .nudge:
                s.lastNudge = event.t
                s.isAsking = true
                if s.config.mode == .slip { s.lastSlip = nil }
            case .down, .through: s.isAsking = false
            }
        }
        state = s
        trace = []
        askedAt = s.lastNudge ?? 0
        route = s.isAsking ? .asking : .session
        detector.threshold = Constants.sensitivityLadder[draft.sensitivity - 1]
        detector.start()
        startClock()
        UIApplication.shared.isIdleTimerDisabled = true
        // A resumed session is still a session: it needs the question scheduled and,
        // if this is the first run, the permission asked for.
        liveActivity.start(task: s.task, armedAt: armedAtDate, endsAt: endsAtDate)
        armNotifications()
    }

    /// The monitor runs in another process and can only leave a timestamp behind.
    /// Fold anything it recorded into the engine the moment the app runs again.
    private func drainMonitorSlip() {
        let recorded = SharedState.lastSlip
        guard recorded > 0, recorded > (state.lastSlip ?? 0) else { return }
        SharedState.lastSlip = 0
        send(.slip, at: recorded)
    }

    /// Asks once, then (re)schedules. Safe to call repeatedly — iOS only shows the
    /// prompt the first time.
    private func armNotifications() {
        Task {
            _ = await notifications.requestAuthorization()
            self.rescheduleQuestions()
        }
    }

    // MARK: - Clock and trace

    private func startClock() {
        stopClock()
        timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.onClock() }
        }
    }

    private func stopClock() {
        timer?.invalidate()
        timer = nil
    }

    private func onClock() {
        nowTick = Date().timeIntervalSince1970
        if nowTick - lastEngineTick >= 1 {
            lastEngineTick = nowTick
            send(.tick, at: nowTick)
        }
    }

    private func sample(_ magnitude: Double) {
        trace.append(magnitude)
        if trace.count > 160 { trace.removeFirst(trace.count - 160) }

        // The live trace is 110 ms samples; the stored one is 5 s buckets, peak held,
        // for the whole-session picture on the closed screen.
        let now = Date().timeIntervalSince1970
        if bucketStart == 0 { bucketStart = now }
        bucketPeak = max(bucketPeak, magnitude)
        if now - bucketStart >= 5 {
            store.appendTrace(bucketPeak)
            bucketPeak = 0
            bucketStart = now
        }
    }
}

/// `CXCallObserver` keeps only a weak delegate, so this is held by the model.
private final class CallWatcher: NSObject, CXCallObserverDelegate {
    private let onChange: @MainActor (Bool) -> Void

    init(onChange: @escaping @MainActor (Bool) -> Void) {
        self.onChange = onChange
    }

    func callObserver(_ observer: CXCallObserver, callChanged call: CXCall) {
        let active = observer.calls.contains { !$0.hasEnded }
        Task { @MainActor in self.onChange(active) }
    }
}

File: android/app/src/main/kotlin/app/ballast/AppModel.kt
package app.ballast

import android.content.Context
import android.content.pm.ApplicationInfo
import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.platform.LocalContext
import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import android.os.Handler
import android.os.Looper

enum class Route { SETUP, SESSION, ASKING, CLOSED }

@Composable fun stringRes(id: Int): String = stringResource(id)
@Composable fun stringRes(id: Int, vararg args: Any): String = stringResource(id, *args)
@Composable fun LocalContextCompat(): Context = LocalContext.current

/** Drives the engine and everything around it. The rules themselves live in Engine.kt. */
class AppModel(private val context: Context) {

    val store = Store(context)
    val route: MutableState<Route> = mutableStateOf(Route.SETUP)
    val now: MutableState<Double> = mutableStateOf(seconds())
    val watched: MutableState<Set<String>> = mutableStateOf(store.prefs.watched.toSet())

    var askedAt = 0.0
        private set
    var closed: SessionRecord? = null
        private set
    private var slipLabel: String? = null

    private var state = SessionState()
    private val ticker = Handler(Looper.getMainLooper())

    init {
        store.open?.let { open ->
            // A session whose deadline has passed should not still be running.
            val deadline = open.startedAt + open.durationMin * 60.0
            if (seconds() > deadline) {
                closed = store.end(deadline)
                Watcher.stop(context)
                route.value = Route.CLOSED
            } else {
                state = rebuild(open)
                route.value = if (state.isAsking) Route.ASKING else Route.SESSION
                Watcher.start(context)
                tick()
            }
        }
    }

    private fun seconds() = System.currentTimeMillis() / 1000.0

    private fun config() = Config(
        intervalMin = Constants.reminderInterval(
            store.prefs.intervalMin, store.prefs.durationMin),
        holdSeconds = store.prefs.holdSeconds)

    private fun rebuild(record: SessionRecord): SessionState {
        var s = SessionState(task = record.task, config = config(),
            sessionStart = record.startedAt)
        record.events.forEach { e ->
            s = when (e.type) {
                RecordKind.SLIP -> s.copy(lastSlip = e.t)
                RecordKind.PICKUP -> s.copy(lastPickup = e.t)
                RecordKind.NUDGE -> s.copy(lastNudge = e.t, isAsking = true)
                RecordKind.DOWN, RecordKind.THROUGH -> s.copy(isAsking = false)
            }
        }
        return s
    }

    fun refresh() {
        watched.value = store.prefs.watched.toSet()
        now.value = seconds()
    }

    // MARK: - Intent

    fun start(task: String, durationMin: Int) {
        val prefs = store.prefs.apply {
            this.task = task
            this.durationMin = durationMin
        }
        store.savePrefs(prefs)
        val t = seconds()
        store.begin(task, durationMin, t)
        send(Event.StartSession(task, config()), t)
        route.value = Route.SESSION
        Watcher.start(context)
        tick()
    }

    fun end() {
        val t = seconds()
        send(Event.EndSession, t)
        closed = store.end(t)
        Watcher.stop(context)
        ticker.removeCallbacksAndMessages(null)
        route.value = Route.CLOSED
    }

    fun slip(packageName: String) {
        if (packageName.isNotEmpty()) {
            store.recordInterruption(packageName)
            slipLabel = label(packageName)
        }
        send(Event.Slip, seconds())
    }

    fun pickup() = send(Event.Pickup, seconds())

    /** The reminder already asked; opening it must never be refused. */
    fun askNow() = send(Event.AskNow, seconds())

    fun dismiss(how: Dismissal) = send(Event.Dismiss(how), seconds())

    fun markDone() {
        closed?.let { store.markCompleted(it.id); closed = store.sessions.firstOrNull { s -> s.id == it.id } }
        route.value = Route.CLOSED
    }

    fun newSession() { closed = null; slipLabel = null; route.value = Route.SETUP }

    fun wipe() { store.wipe(); route.value = Route.SETUP }

    fun toggleWatched(packageName: String) {
        val prefs = store.prefs
        if (!prefs.watched.remove(packageName)) prefs.watched.add(packageName)
        store.savePrefs(prefs)
        watched.value = prefs.watched.toSet()
    }

    // MARK: - Derived

    fun armedIn(atNow: Double): Double? {
        val anchor = state.anchor ?: return null
        val left = state.config.interval - (atNow - anchor)
        return if (left > 0) left else null
    }

    fun askBody(): String {
        val label = slipLabel
        return if (label != null)
            context.getString(R.string.ask_body_slip, label, state.task)
        else {
            val anchor = state.anchor ?: state.sessionStart ?: askedAt
            context.getString(R.string.ask_body, humanDuration(((askedAt - anchor) / 60).toInt()))
        }
    }

    fun label(packageName: String): String = runCatching {
        val pm = context.packageManager
        pm.getApplicationLabel(pm.getApplicationInfo(packageName, 0)).toString()
    }.getOrDefault(packageName)

    /** Only the packages declared in <queries> are visible, which is the point. */
    fun installedCandidates(): List<Pair<String, String>> {
        val pm = context.packageManager
        return CANDIDATES.mapNotNull { pkg ->
            runCatching {
                val info: ApplicationInfo = pm.getApplicationInfo(pkg, 0)
                pkg to pm.getApplicationLabel(info).toString()
            }.getOrNull()
        }.sortedBy { it.second }
    }

    // MARK: - Engine

    private fun send(event: Event, at: Double) {
        val (next, effects) = reduce(state, event, at)
        state = next
        effects.forEach { effect ->
            when (effect) {
                is Effect.Record -> store.append(effect.kind, effect.at)
                is Effect.PresentAsking -> { askedAt = at; now.value = at; route.value = Route.ASKING }
                Effect.PresentClosed -> route.value = Route.CLOSED
                Effect.DismissAsking -> { slipLabel = null; route.value = Route.SESSION }
            }
        }
    }

    private fun tick() {
        ticker.removeCallbacksAndMessages(null)
        ticker.postDelayed(object : Runnable {
            override fun run() {
                now.value = seconds()
                ticker.postDelayed(this, 500)
            }
        }, 500)
    }

    companion object {
        /** Mirrors the <queries> block; the manifest is the source of truth. */
        val CANDIDATES = listOf(
            "com.instagram.android", "com.zhiliaoapp.musically", "com.twitter.android",
            "com.x.android", "com.facebook.katana", "com.google.android.youtube",
            "com.reddit.frontpage", "com.linkedin.android", "com.snapchat.android",
            "com.whatsapp", "org.telegram.messenger", "com.pinterest",
            "com.netflix.mediaclient", "tv.twitch.android.app")
    }
}
